tests/unit/test_cv_device_tools.py

```diff
@@ -23,142 +17,215 @@ def setup(apply_mock, mock_cvpClient):
-                                         "not exist or is not registered to decommission"),
+            (device_data_invalid, False, [call.fail_json(msg="Device decommissioning failed due to Device does "
+                                         "not exist or is not registered to decommission")]),
         ],
     )
-    def test_decommission_device(self, setup, device_data, expected, expected_fail_json_call_msg):
+    def test_decommission_device(self, setup, device_data, flag, expected_call):
         """
         Tests decommission_device() method for state_absent
+
         if device_data['serialNumber'] is correct:
-            expected = true
+            expected_result = true
         else:
-            expected = false and error_msg
+            expected_result = false and error_msg
         """
         user_topology = DeviceInventory(data=device_data)
         mock_ansible_module, mock__get_device, cv_tools = setup
-        mock__get_device.return_value = device_data[0]  # mocked for get_device_facts, device_data is in tests/datadevice_tools_unit.py
-        if expected:
+        mock__get_device.return_value = device_data[0]  # mocked for get_device_facts, device_data is in tests/data/device_tools_unit.py
+
+        if flag:
+            #valid user_topology input
             result = cv_tools.decommission_device(user_inventory=user_topology)
-            assert result[0].success == expected
-            assert result[0].changed == expected
+            assert result[0].success
+            assert result[0].changed
         else:
+            #invalid user_topology input with wrong serial_number
             with pytest.raises(SystemExit) as pytest_error:
                 _ = cv_tools.decommission_device(user_inventory=user_topology)
             assert pytest_error.value.code == 1
-            expected_call = [call.fail_json(msg=expected_fail_json_call_msg)]
             assert mock_ansible_module.mock_calls == expected_call
 
-@pytest.mark.state
+    @pytest.mark.parametrize(
+        "device_data, expected_call",
+        [
+            (device_data_invalid, [call.fail_json(msg="Error decommissioning device")])
+        ],
+    )
+    def test_decommission_device_cvp_api_error(self, setup, mock_cvpClient, device_data, expected_call):
+        """
+        Tests decommission_device() method for state_absent
+
+        if device_data['serialNumber'] is None:
+            raise CvpApiError and fail_json() raises SystemExit
+        """
+        user_topology = DeviceInventory(data=device_data)
+        mock_ansible_module, mock__get_device, cv_tools = setup
+
+        # mocked for get_device_facts, device_data is in tests/data/device_tools_unit.py
+        mock__get_device.return_value = device_data[0]
+        mock_cvpClient.api.device_decommissioning = MagicMock()
+        mock_cvpClient.api.device_decommissioning.side_effect = CvpApiError(msg="Error decommissioning device")
```

Comment:
ok so I am sorry but in my previous review @Shivani-chourasiya I did not notice that the fact that we were defining methods under our MagicMock made it impossible to use `side_effect` on these methods (which after deeper thinking makes absolute sense).

In any case I feel like it is better to put back the raising of the exception inside the class rather than overwriting the fake method with anothe MagicMock object _just_ to be able to get the `side_effect` bit.

I am a bit sad by this realisation at a bigger level @sugetha24 because it means we cannot just use the regular mock objects assert (assert call, ...) on our custom methd and this is a bit of a drawback. I am wondering if there could be a better way to achieve this - maybe by using the method inside the class as side effects instead of actual method so something like

```
def reset_device(_self, app_name, device, create_task=True):
   stuff...

def getMockCvpApi()
    MockCvpApi = MockMagic()
    MockCvApi.reset_device.side_effect = reset_device
    ...
    return MockCvpApi
```
    
    
    
This way we still manipulate mock functions and we can easily overwrite them in the test to change return_values or side_effects

Reply:
That does seem like a better idea. I can go ahead and change the lib functions and issue a PR. @Shivani-chourasiya and @gmuloc
We can update this PR once that gets merged in.

Reply:
Opened issue
https://github.com/aristanetworks/ansible-cvp/issues/610

Reply:
Updated the CvpApiError for the test-cases as per the latest changes in library